core/vm/contracts_write.go

```diff
@@ -150,3 +169,168 @@ func (c *stakingPrecompile) RunWriteCapable(
+	// originating ShardID is pulled from the EVM object, not a parameter
+	crossShardXferABIJSON := `
+	[
+	  {
+	    "inputs": [
+		  {
+	        "internalType": "uint256",
+	        "name": "value",
+	        "type": "uint256"
+	      },
+		  {
+	        "internalType": "address",
+	        "name": "to",
+	        "type": "address"
+	      },
+		  {
+	        "internalType": "uint64",
+	        "name": "toShardID",
+	        "type": "uint32"
+	      }
+	    ],
+	    "name": "crossShardTransfer",
+	    "outputs": [],
+	    "stateMutability": "payable",
+	    "type": "function"
+	  }
+	]`
+	var err error
+	abiCrossShardXfer, err = abi.JSON(strings.NewReader(crossShardXferABIJSON))
+	if err != nil {
+		// means an error in the code
+		panic("Invalid cross shard transfer ABI JSON")
+	}
+}
+
+type crossShardXferPrecompile struct {
+	address common.Address
+}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *crossShardXferPrecompile) RequiredGas(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) (uint64, error) {
+	// we just charge the intrinsic gas here
+	// using data that includes: fromAddress, toAddress, fromShardID, toShardID, value
+	var payload []byte = make([]byte, 0)
+	fromAddress, toAddress, fromShardID, toShardID, value, err := parseCrossShardXferData(evm, contract, input)
+	if err == nil {
+		data := struct {
+			fromAddress common.Address
+			toAddress   common.Address
+			fromShardID uint32
+			toShardID   uint32
+			value       *big.Int
+		}{
+			fromAddress: fromAddress,
+			toAddress:   toAddress,
+			fromShardID: fromShardID,
+			toShardID:   toShardID,
+			value:       value,
+		}
+		if encoded, err := rlp.EncodeToBytes(data); err == nil {
+			payload = encoded
+		}
+	}
+	if gas, err := IntrinsicGas(
+		payload,
+		false,                                   // contractCreation
+		evm.ChainConfig().IsS3(evm.EpochNumber), // homestead
+		evm.ChainConfig().IsIstanbul(evm.EpochNumber), // istanbul
+		false, // isValidatorCreation
+	); err != nil {
+		return 0, err // ErrOutOfGas occurs when gas payable > uint64
+	} else {
+		return gas, nil
+	}
+}
+
+// RunWriteCapable runs the actual contract
+func (c *crossShardXferPrecompile) RunWriteCapable(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+	sentValue *big.Int,
+) ([]byte, error) {
+	fromAddress, toAddress, fromShardID, toShardID, value, err := parseCrossShardXferData(evm, contract, input)
+	if err != nil {
+		return nil, err
+	}
+	// validate not a contract
+	if len(evm.StateDB.GetCode(fromAddress)) > 0 && !evm.IsValidator(evm.StateDB, fromAddress) {
+		return nil, errors.New("cross shard xfer not yet implemented for contracts")
+	}
+	// validate not a contract
+	if len(evm.StateDB.GetCode(toAddress)) > 0 && !evm.IsValidator(evm.StateDB, toAddress) {
```

Comment:
We don't have the restriction on normal cross-shard txn that the destination address can't be a contract. it can be any address for toAddress. fromAddress has to be EOA address since it's sent from EOA address.